feat: Add per model metrics

VedantMahabaleshwarkar opened the pull request:
#### Motivation

#60 

#### Modifications

- Added `modelId` parameter to `logTimingMetricDuration` function in `Metrics.java`
    - This adds the `modelId` label to the following metrics : `modelmesh_cache_miss_milliseconds`, `modelmesh_loadmodel_milliseconds`, `modelmesh_unloadmodel_milliseconds`, `modelmesh_req_queue_delay_milliseconds`, `modelmesh_model_sizing_milliseconds`, `modelmesh_age_at_eviction_milliseconds`
- Added `modelId` parameter to `logSizeEventMetric` function in `Metrics.java`
    - This adds the `modelId` label to the following metrics : `modelmesh_loading_queue_delay_milliseconds`, `modelmesh_loaded_model_size_bytes`
- Added `modelId` and `vmodelId` parameter to `logRequestMetrics` function in `Metrics.java`
    - This adds the modelId/vmodelId (whichever applicable) to the following metrics: `modelmesh_invoke_model_milliseconds`, `modelmesh_api_request_milliseconds`

#### Result

Metrics listed above have the modelid label attached to the emitted metric

rafvasq commented:
Hey @VedantMahabaleshwarkar, just wondering if this is replacing #86? If so, maybe that one can be closed.

VedantMahabaleshwarkar commented:
> Hey @VedantMahabaleshwarkar, just wondering if this is replacing #86? If so, maybe that one can be closed.

@rafvasq Yep this PR will be replacing the old one, I'll close it out.

njhill commented on src/main/java/com/ibm/watson/modelmesh/Metrics.java:
```diff
@@ -220,10 +237,15 @@ public PrometheusMetrics(Map<String, String> params, Map<String, String> infoMet
                 if (m == API_REQUEST_TIME || m == API_REQUEST_COUNT || m == INVOKE_MODEL_TIME
-                    || m == INVOKE_MODEL_COUNT || m == REQUEST_PAYLOAD_SIZE || m == RESPONSE_PAYLOAD_SIZE) {
-                    builder.labelNames("method", "code");
+                        || m == INVOKE_MODEL_COUNT || m == REQUEST_PAYLOAD_SIZE || m == RESPONSE_PAYLOAD_SIZE) {
+                    if (this.enablePerModelMetrics && m.type != COUNTER_WITH_HISTO) {
+                        builder.labelNames("method", "code", "modelId");
+                    } else {
+                        builder.labelNames("method", "code");
+                    }
+                } else if (this.enablePerModelMetrics && m.type != GAUGE && m.type != COUNTER && m.type != COUNTER_WITH_HISTO) {
+                    builder.labelNames("modelId");
```
I think we should add the `modelId` label based on a specific list of metric names rather than the metric types. Since in general there's no reason why a particular type should be model-specific or not.

VedantMahabaleshwarkar replied:
we are adding the label on a list of metrics 
```
if (m == API_REQUEST_TIME || m == API_REQUEST_COUNT || m == INVOKE_MODEL_TIME || m == INVOKE_MODEL_COUNT || m == REQUEST_PAYLOAD_SIZE || m == RESPONSE_PAYLOAD_SIZE)
```
I removed the check for `m.type != COUNTER_WITH_HISTO`

njhill commented on src/main/java/com/ibm/watson/modelmesh/Metrics.java:
```diff
@@ -342,13 +368,21 @@ public void logTimingMetricSince(Metric metric, long prevTime, boolean isNano) {
-        public void logTimingMetricDuration(Metric metric, long elapsed, boolean isNano) {
-            ((Histogram) metricsMap.get(metric)).observe(isNano ? elapsed / M : elapsed);
+        public void logTimingMetricDuration(Metric metric, long elapsed, boolean isNano, String modelId) {
+            if (enablePerModelMetrics) {
```
We could accept a `null` `modelId` here so that it can be omitted for metrics where it isn't applicable. The check would then be `if (enablePerModelMetrics && modelId != null)`

VedantMahabaleshwarkar replied:
fixed

njhill commented on src/main/java/com/ibm/watson/modelmesh/ModelMeshApi.java:
```diff
@@ -87,6 +87,7 @@
 import io.netty.util.ReferenceCountUtil;
 import io.netty.util.concurrent.FastThreadLocalThread;
 import org.apache.thrift.TException;
+import org.checkerframework.checker.units.qual.A;
```
Looks like this was included accidentally

VedantMahabaleshwarkar replied:
removed

njhill commented on src/main/java/com/ibm/watson/modelmesh/ModelMesh.java:
```diff
@@ -3989,9 +3989,10 @@ else if (mr.getInstanceIds().containsKey(instanceId)) {
             throw t;
         } finally {
             if (methodStartNanos > 0L && metrics.isEnabled()) {
+                String[] extraLabels = new String[]{modelId};
```
Doesn't look like this is used?

VedantMahabaleshwarkar replied:
removed

njhill commented on src/main/java/com/ibm/watson/modelmesh/ModelMeshApi.java:
```diff
@@ -783,8 +784,13 @@ public void onHalfClose() {
-                                status.getCode(), reqSize, respSize);
+                        if (isVModel) {
+                            metrics.logRequestMetrics(true, methodName, nanoTime() - startNanos,
+                                    status.getCode(), reqSize, respSize, "", Iterables.toString(modelIds));
```
`Iterables.toString()` I think will produce a comma-separated list of the contained strings. That's not what we want for the label. Passing mutliple model or vmodel ids is kind of a niche case but really we should iterate over them and log separate metrics for each.

VedantMahabaleshwarkar replied:
resolved, looping over the list and logging metrics separately

njhill reviewed (commented):
Thanks @VedantMahabaleshwarkar, and apologies for taking so long to look at this.

It looks like a good start but I think some more work is needed around how modelIds / vmodelIds are handled.

Currently, in the `logRequestMetrics` method, you added both modelId and vmodelId args, but only ever pass in one of those.. and inside the method only choose the non-null one. So it may as well just take one arg.

However I think we should aim to always log the model id, and additionally log the vmodel id too whenever it's used.

Requests can be targeted directly at concrete models (modelId) and in this case it's pretty straightforward and the changes you've made should cover it.

But when they are targeted at a vmodelId, we should:
- Include a vmodelId for all the same metrics that we include modelId - this will require new changes to propagate the vmodelId with the request because currently it's "lost" below the top of the stack, once it's been resolved to a modelId.
- Also include the resolved modelId *everywhere* for all these same metrics. This is already done in all these places lower in the stack, but the top level logRequestMetrics in the ModelMeshApi class will only have the vModelId and so we'll need to make sure the *resolved* modelId is obtained in this same place and included as a label alongside the vmodelId.

For propagating the vMmodelId, I would suggest adding it to the Litelinks ThreadContext ... this is a ThreadLocal map which will be automatically be transferred between modelmesh pods if the resquest if forwarded.

I hope the above makes some sense, feel free to ping if not!

VedantMahabaleshwarkar commented:
@njhill Implemented all of your suggestions. 
When a request is targeted for a vmodelid, the resolved modelid is also included as a label. According to your conversation I have added the vmodelid to the threadcontext to propogate it before we resolve it to the modelid. 
The metrics with these new labels will always have both labels, but with an empty string value if the label is not applicable. (This is because you can't pass in a null value as a label)

VedantMahabaleshwarkar commented:
/test all

njhill commented on src/main/java/com/ibm/watson/modelmesh/ModelMeshApi.java:
```diff
@@ -784,12 +788,25 @@ public void onHalfClose() {
-                        } else {
-                            metrics.logRequestMetrics(true, methodName, nanoTime() - startNanos,
-                                    status.getCode(), reqSize, respSize, Iterables.toString(modelIds), "");
+                        Iterator<String> midIt = modelIds.iterator();
+                        while (midIt.hasNext()) {
+                            if (isVModel) {
+                                String mId = null;
+                                String vmId = midIt.next();
+                                try {
+                                    mId = vmm().resolveVModelId(midIt.next(), mId);
+                                    metrics.logRequestMetrics(true, methodName, nanoTime() - startNanos,
+                                    status.getCode(), reqSize, respSize, mId, vmId);
+                                }
+                                catch (Exception e) {
+                                    logger.error("Could not resolve model id for vModelId" + vmId, e);
+                                    metrics.logRequestMetrics(true, methodName, nanoTime() - startNanos,
+                                    status.getCode(), reqSize, respSize, "", vmId);
+                                }
+                            } else {
+                                metrics.logRequestMetrics(true, methodName, nanoTime() - startNanos,
+                                    status.getCode(), reqSize, respSize, midIt.next(), "");
+                            }
```
@VedantMahabaleshwarkar thinking more here, it's probably too complicated to log the metrics in the multi-id case. This functionality was for a fairly unique use and I doubt it's actually being used by anyone anymore.

So I would say adjust the logic further up in this method to ensure the modelId / vModelId variables are set by the time it reaches this finally block.

I.e. add a local vModel variable and ensure that this is set in the vModel case. Maybe add a resolvedModelId variable too, which can be set to the modelId above in the modelId case and otherwise set within this finally block above, from the new ThreadLocal mentioned in the other comment (prior to the call to `clearThreadLocals()`). I'm suggesting to call it resolvedModelId rather than modelId since the latter currently refers to ether modelId or vModelId.

Maybe in the multi-id case we leave both these vars as null and then the logic here will know to just not record any metric at all.

njhill replied:
@VedantMahabaleshwarkar WDYT about this suggestion?

VedantMahabaleshwarkar replied:
@njhill does the change look good to you? I separated out the single-model and multi-model case and the rest of the logic is in line with your suggestion I think

njhill reviewed (commented):
Thanks @VedantMahabaleshwarkar, added some more comments.

VedantMahabaleshwarkar commented:
@njhill addressed most of your comments and avoided passing around the `boolean isVModel`. (@danielezonca helped :) ) 
The only thing we didn't quite get was your suggestion to use `FastThreadLocal` to store the `vModelId`. It seemed redundant given that we are already adding it to the `ThreadContext`? Other than that every suggestion has been implemented. 

Other than that @danielezonca also had a small suggestion to change the `VModelManager` Class. Let me know if it should be kept for a separate PR.

rafvasq commented on src/main/java/com/ibm/watson/modelmesh/ModelMesh.java:
```diff
@@ -3315,6 +3315,7 @@ protected Map<String, ServiceInstanceInfo> getMap(Object[] arr) {
     static final String KNOWN_SIZE_CXT_KEY = "tas.known_size";
     static final String UNBALANCED_KEY = "mmesh.unbalanced";
     static final String DEST_INST_ID_KEY = "tas.dest_iid";
+    static final String VMODELID = "vmodelid";
```
nit: Naming consistency 

```suggestion
    static final String VMODEL_ID = "vmodelid";
```

ckadner replied:
I agree with Rafael's nit :-)

rafvasq reviewed (commented):
While I can't speak to the [the use of `FastThreadLocal`](https://github.com/kserve/modelmesh/pull/90#issuecomment-1619116491), in terms of usability, I deployed the changes myself and was able to filter metrics by model using the new parameter `modelId`. I'd also recommend saving the changes to the `VModelManager` class for a separate PR if it's not directly required to this one. 

FYI @ckadner

ckadner commented on src/main/java/com/ibm/watson/modelmesh/Metrics.java:
```diff
@@ -56,14 +61,14 @@
 
     boolean isEnabled();
-
     void logTimingMetricSince(Metric metric, long prevTime, boolean isNano);
```
nit: why remove empty line above?

ckadner reviewed (changes requested):
I have a few more nit-picks. Beyond those it looks like there are a few more unresolved conversations, beyond the "FastThreadLocal" topic.

njhill reviewed (commented):
Thanks @VedantMahabaleshwarkar @danielezonca and apologies for the delay, I was out last week and still catching up.

> The only thing we didn't quite get was your suggestion to use `FastThreadLocal` to store the `vModelId`. It seemed redundant given that we are already adding it to the `ThreadContext`?

I was referring to the resolved `modelId` here, not the `vModelId`. And it's just for the purpose of "passing it out" of this method so that it can be included in the metrics recorded in the calling function in addition to the vModelId.

VedantMahabaleshwarkar commented on src/main/java/com/ibm/watson/modelmesh/ModelMeshApi.java:
```diff
@@ -430,20 +434,22 @@ public void ensureLoaded(EnsureLoadedRequest request, StreamObserver<ModelStatus
-        modelId = null;
         boolean first = true;
         while (true) {
-            modelId = vmm().resolveVModelId(vModelId, modelId);
```
Removed the `originalModelId` change as a part of this revision

ckadner commented:
Hi Vedant, since you force-pushed the last commit, I cannot find where/whether Nicks suggestion made it into your last code changes. Are the remaining 7 unresolved conversations actually resolved? Also, you will need to update your branch once more :-)

VedantMahabaleshwarkar commented:
@njhill pulled in your changes into this PR

njhill reviewed (approved):
Thanks @VedantMahabaleshwarkar for all of your patience with this!
